rust/connlib/tunnel/src/dns.rs

```diff
@@ -161,3 +162,66 @@ where
     }
     Some(answer_builder.finish())
 }
+
+fn resource_from_question<N: ToDname>(
+    resources: &ResourceTable<ResourceDescription>,
+    question: &Question<N>,
+) -> Option<ResolveStrategy<ResourceDescription, DnsQueryParams>> {
+    let name = ToDname::to_cow(question.qname()).to_string();
+    let qtype = question.qtype();
+    match qtype {
+        Rtype::A | Rtype::Aaaa => resources
+            .get_by_name(&name)
+            .cloned()
+            .map(ResolveStrategy::LocalResponse)
+            .unwrap_or(ResolveStrategy::new(name, qtype))
+            .into(),
+        Rtype::Ptr => {
+            let mut dns_parts = name.split('.').rev();
+            if !dns_parts
+                .next()
+                .is_some_and(|d| d == REVERSE_DNS_ADDRESS_END)
+            {
+                return None;
+            }
+            let ip: IpAddr = match dns_parts.next() {
+                Some(REVERSE_DNS_ADDRESS_V4) => {
+                    let mut ip = [0u8; 4];
+                    for i in ip.iter_mut() {
+                        *i = dns_parts.next()?.parse().ok()?;
+                    }
+                    ip.into()
+                }
+                Some(REVERSE_DNS_ADDRESS_V6) => {
+                    let mut ip = [0u8; 16];
+                    for i in ip.iter_mut() {
+                        *i = u8::from_str_radix(
+                            &format!("{}{}", dns_parts.next()?, dns_parts.next()?),
+                            16,
+                        )
+                        .ok()?;
+                    }
+                    ip.into()
+                }
+                _ => return None,
+            };
+
+            if dns_parts.next().is_some() {
+                return None;
+            }
+
+            resources
+                .get_by_ip(ip)
+                .cloned()
+                .map(ResolveStrategy::LocalResponse)
+                .unwrap_or(ResolveStrategy::new(name, qtype))
+                .into()
+        }
+        _ => None,
+    }
```

Comment:
This might just be remnants from my full-stack engineering days, but would it make sense to condense some of this to functions we can unit test more easily later?

Reply:
Good call, I'll move the reverse DNS extraction into a separated function. It'll make things more readable. It kind of seems like black magic as is.